test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+        subarr_idxs[i] =
+            dims[i] > 1 ? af::seq(pad_size, pad_size + dims[i] - 1) : af::span;
+    }
+
+    metadata.initArrays(full_arr_dims, ty, subarr_idxs);
+}
+
+// Generates a reordered array. testWriteToOutputArray expects that this array
+// will still have the correct output values from the af_* function, even though
+// the array was initially reordered.
+void genReorderedArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    // This reorder combination will not move data around, but will simply
+    // call modDims and modStrides (see src/api/c/reorder.cpp).
+    // The output will be checked if it is still correct even with the
+    // modified dims and strides "hack" with no data movement
+    unsigned reorder_0 = 0;
+    unsigned reorder_1 = 2;
+    unsigned reorder_2 = 1;
+
+    // Shape the output array such that the reordered output array will have
+    // the correct dimensions that the test asks for (i.e. must match dims arg)
+    af::dim4 out_dims(dims[reorder_0], dims[reorder_1], dims[reorder_2]);
+    metadata.initArrays(out_dims, ty);
+    metadata.setOutputArray(af::reorder(metadata.out_arr,
+                                        reorder_0, reorder_1, reorder_2));
+}
+
+// Partner function of testWriteToOutputArray. This generates the "special"
+// array that testWriteToOutputArray will use to check if the af_* function
+// correctly uses an existing array as its output
+void genTestOutputArray(af_array *out_ptr,
+                        const unsigned ndims, const dim_t *dims,
+                        const af::dtype ty,
+                        TestOutputArrayInfo* metadata) {
+    af::dim4 arr_dims(ndims, dims);
+    switch (metadata->getOutputArrayType()) {
+    case NULL_ARRAY:
+        genNullArray(*metadata, arr_dims, ty);
+        break;
+    case FULL_ARRAY:
+        genRegularArray(*metadata, arr_dims, ty);
+        break;
+    case SUB_ARRAY:
+        genSubArray(*metadata, arr_dims, ty);
+        break;
+    case REORDERED_ARRAY:
+        genReorderedArray(*metadata, arr_dims, ty);
+        break;
+    }
+    *out_ptr = metadata->getOutputArrayC();
+}
+
+// Partner function of genTestOutputArray. This uses the same "special"
+// array that genTestOutputArray generates, and checks whether the
+// af_* function wrote to that array correctly
+::testing::AssertionResult
+testWriteToOutputArray(std::string gold_name, std::string result_name,
+                       const af_array gold, const af_array out,
+                       TestOutputArrayInfo *metadata) {
+    // In the case of NULL_ARRAY, the output array starts out as null.
+    // After the af_* function is called, it shouldn't be null anymore
+    if (metadata->getOutputArrayType() == NULL_ARRAY) {
+        if (out == 0) {
+            return ::testing::AssertionFailure()
+                << "Output af_array " << result_name << " is null";
+        }
+        metadata->setOutputArrayC(out);
```

Comment:
will become
`metadata->setOutputArray(af::array())`

Reply:
Actually I don't think this line will make any difference, even in my original code. In the `NULL_ARRAY` case, we only check two things - that the output of `af_approx1` isn't null, and that the output matches the reference, element-wise. We do the latter in the very last else statement in this function (`testWriteToOutputArray`), and in there we just use the arguments given in the call, not whatever data we've stored in `TestOutputArrayInfo` (it'll be a different story for `SUB_ARRAY`, however).